Runtime/Scripts/DataCollection/RGDataCollection.cs

```diff
@@ -124,135 +152,150 @@ public async Task SaveReplayDataInfo(long clientId, RGStateActionReplayData repl
-                
-                // Convert the list into JSONL format
-                List<string> jsonLines = new List<string>();
-                foreach (var validation in validations)
-                {
-                    string jsonString = JsonConvert.SerializeObject(validation);
-                    jsonLines.Add(jsonString);
-                }
+                state.validationSummary.passed += validations.Count(rd => rd.result == RGValidationResultType.PASS);
+                state.validationSummary.failed += validations.Count(rd => rd.result == RGValidationResultType.FAIL);
+                state.validationSummary.warnings += validations.Count(rd => rd.result == RGValidationResultType.WARNING);
+
+                var validationsLinesString = SerializeJsonLines(validations);
 
-                // Combine the JSON strings with newline characters
-                string jsonLinesString = string.Join("\n", jsonLines) + "\n";
-                
                 // Wait for any prior validation write task to finish
-                if (_clientIdToValidationDataTask.TryRemove(clientId, out var validationDataTask))
+                if (state.validationDataTask != null)
                 {
-                    await validationDataTask;
+                    await state.validationDataTask;
+                    state.validationDataTask = null;
                 }
 
                 // Write out the validations to the JSONL file; note.. the prior write mush finish before the next tick
                 // otherwise we can't write to the file safely
                 var validationFilePath =
                     GetSessionDirectory($"validationData/{clientId}/rgbot_validations.jsonl");
-                var validationTask = File.AppendAllTextAsync(validationFilePath, jsonLinesString);
-                _clientIdToValidationDataTask[clientId] = validationTask;
+                var validationTask = File.AppendAllTextAsync(validationFilePath, validationsLinesString);
+                state.validationDataTask = validationTask;
+            }
+
+            // Atomically swap the logs queue with an empty queue.
+            // Then the log collector can upload this queue's logs while logs continue to accumulate in the new queue
+            var logs = Interlocked.Exchange(ref state.logs, new());
+            if (!logs.IsEmpty)
+            {
+                state.logFlushTask = FlushLogs(clientId, logs);
             }
-            
+
             // remove any already completed tasks from the tracker
-            replayDataTasks.RemoveAll(v => v.IsCompleted);
+            state.replayDataTasks.RemoveAll(v => v.IsCompleted);
 
         }
 
         public async Task SaveBotInstanceHistory(long clientId)
         {
-            try
+            // Removing the state from the dictionary will prevent any further data from being saved
+            var state = _clientIdToState.TryRemove(clientId, out var s) ? s : null;
```

Comment:
This is intentionally done outside the `try` because there's no reason it _should_ throw and the `Cleanup` in the `finally` block assumes it's been done already.